anyblock-offshift/adapter.js

```diff
@@ -0,0 +1,139 @@
+const { Requester, Validator } = require('@chainlink/external-adapter')
+
+const customError = (data) => {
+  return !data.hits || !data.hits.hits || data.hits.hits.length < 1
+}
+
+const customParams = {
+  debug: false,
+  roundDay: false,
+  start: false,
+  end: false
+}
+
+const headers = {
+  'Content-Type': 'application/json',
+  Authorization: `Bearer ${process.env.API_KEY}`
+}
+
+function buildVWAP (response, debug) {
+  const sources = response.data.hits.hits.map(i => {
+    const reserve0 = i._source.args.find(j => j.pos === 0)
+    const reserve1 = i._source.args.find(j => j.pos === 1)
+    return {
+      timestamp: i._source.timestamp,
+      reserve0: parseInt(reserve0['value.hex'], 16),
+      reserve1: parseInt(reserve1['value.hex'], 16)
+    }
+  })
+
+  let overallVolume = 0
+  let sumAmountAndPrices = 0
+  for (let i = 1; i < sources.length; i++) {
+    const reserve0volume = Math.abs(sources[i].reserve0 - sources[i - 1].reserve0)
+    const price = sources[i].reserve0 / sources[i].reserve1
+    // reserve1volume: Math.abs(sources[i].reserve1 - sources[i - 1].reserve1),
+    overallVolume += reserve0volume
+    sumAmountAndPrices += price * reserve0volume
+  }
+
+  const r = {
+    status: response.status,
+    data: {
+      result: {
+        volume: overallVolume,
+        vwap: sumAmountAndPrices / overallVolume
+      }
+    }
+  }
+
+  if (debug) {
+    r.data.raw = sources
+  }
+  return r
+}
+
+const createRequest = (input, callback) => {
+  const validator = new Validator(callback, input, customParams)
+  const jobRunID = validator.validated.id
+  const debug = validator.validated.data.debug || false
+  const roundDay = validator.validated.data.roundDay || false
+  let start = validator.validated.data.start
+  let end = validator.validated.data.end
+  const url = 'https://api.anyblock.tools/ethereum/ethereum/mainnet/es/event/search/'
+
+  try {
+    end = parseInt(end)
+    if (roundDay) {
+      const date = new Date(end)
+      date.setUTCHours(0, 0, 0, 0)
+      end = date.getTime()
+    }
+  } catch (err) {
+  }
+
+  try {
+    start = parseInt(start)
+    if (roundDay) {
+      const date = new Date(start)
+      date.setUTCHours(0, 0, 0, 0)
+      start = date.getTime()
+    }
+  } catch (err) {
+  }
+
+  if (!start && !end) {
+    const date = new Date()
+    date.setUTCHours(0, 0, 0, 0)
+    end = date.getTime() / 1000
+
+    start = end - 60 * 60 * 24
+  } else if (!start) {
+    start = end - 60 * 60 * 24
+  } else if (!end) {
+    end = start + 60 * 60 * 24
+  } else if (start === end) {
+    start = end - 60 * 60 * 24
+  }
+
+  const body = {
+    query: {
+      bool: {
+        filter: [
+          { term: { 'address.raw': '0x2B9e92A5B6e69Db9fEdC47a4C656C9395e8a26d2' } },
+          { term: { 'event.raw': 'Sync' } },
```

Comment:
It would be awesome if we could make this more flexible, so we can get VWAP from other AMM protocols like [Balancer](https://github.com/balancer-labs/balancer-core/blob/f4ed5d65362a8d6cec21662fb6eae233b0babc1f/contracts/BPool.sol#L28) for example.

To do this we would need to parameterize:

- network
- address
- event + params

Please share your thoughts on this.

Reply:
Mhhh, good point. It looks like we need some small abstraction of our API - what about an own module for this?

I don't even think that the parameters must be parameterized. Event, address and network should be sufficient. 
The caller of the module should/will know which parameters to be expected.
I'd say a timestamp window would be needed as well.

The current result is a list of events like this:
```
                    [
                        {
                            "value.hex": "0x1b287a18a87a333775bc",
                            "pos": 0,
                            "name": "reserve0",
                            "value.type": "uint112"
                        },
                        {
                            "value.hex": "0x16ad6bac17b7bad757",
                            "pos": 1,
                            "name": "reserve1",
                            "value.type": "uint112"
                        }
                    ]
```